src/jupyterlabcontroller/storage/docker.py

```diff
@@ -1,202 +1,313 @@
-
-    async def list_tags(self, authenticate: bool = True) -> List[str]:
-        """List all the tags.
-
-        Lists all the tags for the repository this client is used with.
+    async def list_tags(self, registry: str, repository: str) -> list[str]:
+        """List all the tags for a given registry and repository.
 
         Parameters
         ----------
-        authenticate: should we try and authenticate?  Used internally
-          for retrying after successful authentication.
+        registry
+            Hostname of Docker container registry.
+        repository
+            Repository of images (for example, ``lsstsqre/sciplat-lab``).
+
+        Returns
+        -------
+        list of str
+            All the tags found for that repository.
         """
-        url = f"https://{self.host}/v2/{self.repository}/tags/list"
-        r = await self.http_client.get(url, headers=self.headers)
-        if r.status_code == 200:
-            body = r.json()
-            return body["tags"]
-        elif r.status_code == 401 and authenticate:
-            await self._authenticate(r)
-            return await self.list_tags(authenticate=False)
-        else:
-            msg = f"Unknown error listing tags from <{url}>: {r}"
-            raise DockerRegistryError(msg)
+        url = f"https://{registry}/v2/{repository}/tags/list"
+        r = await self._client.get(url, headers=self._build_headers(registry))
+        if r.status_code == 401:
+            headers = await self._authenticate(registry, r)
+            r = await self._client.get(url, headers=headers)
+        try:
+            r.raise_for_status()
+            return r.json()["tags"]
+        except Exception as e:
+            msg = f"Error listing tags from <{url}>"
+            raise DockerRegistryError(msg) from e
 
     async def get_image_digest(
-        self, tag: str, authenticate: bool = True
+        self, registry: str, repository: str, tag: str
     ) -> str:
-        """Get the digest of a tag.
-
-        Get the digest associated with an image tag.
+        """Get the digest associated with an image tag.
 
         Parameters
         ----------
-        tag: the tag to inspect
-        authenticate: should we to authenticate?  Used internally for
-          retrying after successful authentication.
+        registry
+            Hostname of Docker container registry.
+        repository
+            Repository of images (for example, ``lsstsqre/sciplat-lab``).
+        tag
+            The tag to inspect.
 
-        Returns the digest as a string, such as "sha256:abcdef"
+        Returns
+        -------
+        str
+            The digest, such as ``sha256:abcdef``.
         """
-        url = f"https://{self.host}/v2/{self.repository}/manifests/{tag}"
-        r = await self.http_client.head(url, headers=self.headers)
-        if r.status_code == 200:
+        url = f"https://{registry}/v2/{repository}/manifests/{tag}"
```

Comment:
I eagerly await the day T&S needs to do an HTTP rather than an HTTPS repository because reasons.